server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java

```diff
@@ -180,31 +185,116 @@ public State reduce(RamAccounting ramAccounting, State state1, State state2) {
     }
 
     @Override
-    public List<Object> terminatePartial(RamAccounting ramAccounting, State state) {
-        if (state instanceof TopKState topKState) {
-            ItemsSketch.Row<Object>[] frequentItems = topKState.itemsSketch.getFrequentItems(ErrorType.NO_FALSE_NEGATIVES);
-            int limit = Math.min(frequentItems.length, topKState.limit);
-            var result = new ArrayList<>(limit);
-            for (int i = 0; i < limit; i++) {
-                var item = frequentItems[i];
-                result.add(Map.of("item", item.getItem(), "frequency", item.getEstimate()));
-            }
-            return result;
-        }
-        return List.of();
+    public List<Map<String, Object>> terminatePartial(RamAccounting ramAccounting, State state) {
+        return state.result();
     }
 
     public DataType<?> partialType() {
         return new StateType(boundSignature.argTypes().getFirst());
     }
 
+    @Nullable
+    @Override
+    public DocValueAggregator<?> getDocValueAggregator(LuceneReferenceResolver referenceResolver,
+                                                       List<Reference> aggregationReferences,
+                                                       DocTableInfo table,
+                                                       List<Literal<?>> optionalParams) {
+        if (aggregationReferences.size() == 1) {
+            Reference reference = aggregationReferences.getFirst();
+            if (reference.hasDocValues()) {
+                if (optionalParams.isEmpty()) {
+                    return getDocValueAggregator(reference, DEFAULT_LIMIT);
+                } else if (optionalParams.size() == 1) {
+                    DataType<?> signatureType = boundSignature.argTypes().getFirst();
+                    Literal<?> parameter = optionalParams.getFirst();
+                    DataType<?> paramType = parameter.valueType();
+                    DataType<?> referenceType = reference.valueType();
+                    if (signatureType.id() != referenceType.id() || paramType.id() != DataTypes.INTEGER.id()) {
+                        throw new IllegalArgumentException("Invalid parameters for topk aggregation.");
+                    }
```

Comment:
Given that the signature matching should make sure that the second parameter fits an integer this either shouldn't happen, or turns a valid case into an error - which mustn't happen due to an optimization.
This this could probably return `null` to fallback to the default implementation but imho going through the signature types with comparisons to true to figure out if the reference is for the first column looks like a hack.

Seems to me that we should adapt the `getDocValueAggregator` interface to make it easier to correlate the reference to the parameter.

This could also be done upfront on master because on a quick glance it looks like `CmpByAggregation` has a similar problem.

Reply:
The simplistic solution would be to also handle in List<Symbol> inputs of the aggregation into `getDocValueAggregator`.